File: src/bin/0153-find-minimum-in-rotated-sorted-array.cpp
// @leetup=info id=153 lang=cpp slug=find-minimum-in-rotated-sorted-array

#include <algorithm>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  int findMin(std::vector<int>& nums) {
    if (nums.size() == 1 || nums.front() < nums.back()) {
      return nums.front();
    }
    return *std::lower_bound(nums.begin(), nums.end(), nums.front(),
                             [](int n, int front) { return front <= n; });
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0153-find-minimum-in-rotated-sorted-array.py
# @leetup=info id=153 lang=python3 slug=find-minimum-in-rotated-sorted-array



# @leetup=code
class Solution:
    def findMin(self, nums: list[int]) -> int:
        N = len(nums)
        if N == 1:
            return nums[0]
        if nums[0] < nums[-1]:
            # Not rotated
            return nums[0]
        left = 0
        right = N

        while left < right:
            mid = (left + right) // 2
            if nums[0] <= nums[mid]:
                left = mid + 1
            else:
                right = mid

        return nums[left]


# @leetup=code
